materialize-bigquery/transactor.go

```diff
@@ -2,287 +2,192 @@ package main
-	// append the SQL for that table.
-	var edcTableDefs = make(map[string]bigquery.ExternalData)
-	for _, b := range t.bindings {
-
-		if b.store.mergeFile != nil {
-			if err := b.store.mergeFile.Close(); err != nil {
-				return fmt.Errorf("mergefile close: %w", err)
-			}
-
-			// Setup the tempTableName pointing to our cloud storage external table definition.
-			edcTableDefs[b.store.tempTableName] = b.store.extDataConfig
-
-			subqueries = append(subqueries, b.store.sql)
-
-			// Clean up the temporary files when store complete (or we error out).
-			defer func(defb *binding) {
-				if err := defb.store.mergeFile.Delete(ctx); err != nil {
-					log.Errorf("could not delete store mergefile: %v", err)
-				}
-				defb.store.mergeFile = nil
-			}(b) // b is part of for loop, make sure you're referencing the right b.
+	for _, binding := range t.bindings {
+		name, err := randomString()
+		if err != nil {
+			return pf.DriverCheckpoint{}, err
 		}
-	}
-
-	// Complete the transaction and return the appropriate error.
-	subqueries = append(subqueries, `
-	COMMIT TRANSACTION;
 
-	SELECT '' AS error;
+		binding.Reset(ctx, fmt.Sprintf("%s/%s", t.config.BucketPath, name))
 
-	EXCEPTION WHEN ERROR THEN
-		ROLLBACK TRANSACTION;
-		SELECT @@error.message AS error;
-	END;
-	`)
-
-	// Build the bigquery query of the combined subqueries.
-	query := t.ep.newQuery(strings.Join(subqueries, "\n"), args...)
-	query.TableDefinitions = edcTableDefs // Tell the query where to get the external references in gcs.
+		t.checkpoint.Bindings = append(t.checkpoint.Bindings, &DriverCheckPointBinding{
+			FilePath: binding.FilePath(),
+			Query:    binding.Queries.WriteSQL,
+		})
+	}
 
-	// This returns a single row with the error status of the query.
-	job, err := t.ep.runQuery(ctx, query)
+	jsn, err := json.Marshal(t.checkpoint)
 	if err != nil {
-		return fmt.Errorf("load query: %w", err)
+		return pf.DriverCheckpoint{}, fmt.Errorf("creating checkpoint json: %w", err)
 	}
 
-	// Fetch te query status from the transaction
-	var queryStatus struct {
-		Error string `bigquery:"error"`
-	}
+	return pf.DriverCheckpoint{
+		DriverCheckpointJson: jsn,
+	}, nil
+}
 
-	if err := t.ep.fetchOne(ctx, job, &queryStatus); err != nil {
-		return fmt.Errorf("fetch one: %w", err)
+func (t *transactor) Store(it *pm.StoreIterator) error {
+	for it.Next() {
+		binding := t.bindings[it.Binding]
+		binding.Store(binding.GenerateDocument(it.Key, it.Values, it.RawJSON))
 	}
 
-	if queryStatus.Error != "" {
-		return fmt.Errorf("merge error: %s", queryStatus.Error)
+	return nil
+}
+
+func (t *transactor) Commit(ctx context.Context) error {
+	for _, binding := range t.bindings {
+		err := binding.Commit(ctx)
+		if err != nil {
+			return err
+		}
 	}
 
 	return nil
 }
 
-func (t *transactor) Acknowledge(context.Context) error {
+func (t *transactor) Acknowledge(ctx context.Context) error {
+	for i, cpBinding := range t.checkpoint.Bindings {
```

Comment:
I think the idea behind the protocol is that since Commit happens before Acknowledge, during the Commit phase you must make sure that the data is committed properly, and if so,  you do not need to run checks here about whether the data is committed or not. Acknowledge can assume that the files must have been written and comitted successfuly.

Reply:
I think most of the questions you have revolves around the writer{} struct on how it's used. So, I'll try to wrap my answer here to includes all of your other comments. First, my understanding is that the Acknowledge phase can come from 1 of two places:

- After the the stream is opened and `RunTransactor` is called, which would happen when a restart happened between Commit and Acknowledged, and so, the bindings needs to be properly configured to execute a query that includes the cloud storage object and the query.
- After a Commit phase, which is the normal operation flow. In this case, the data is being set up in the Prepare phase, Committed to cloud storage in the Commit phase, and then, everything is packaged and ready to be processed in the Acknowledge phase

So, if that first scenario is indeed possible, it means that the Acknowledged phase is going to attempt to create a query for a file path in Google Storage that we're not 100% sure at this point exists. This could lead to a scenario where the connector tries to execute a query for a file that does not exists and fails in a loop forever since we never go pass that driver checkpoint (loads the connector -> acknowledge failed -> connector shuts down -> loads the connector, etc.)

The reason why I put the `writer{}` inside the Binding object is that, as far as I understand, we either need to store the Google cloud object handle or need to reinitialize it on every phase that requires it. In the current way I got it setup, it would mean that we'd setup those objects in Store/Commit/Acknowledge at the very least.

By having a "mutable" sub-object in `writer` it allowed me to carry some state over the pipeline. Honestly, if it causes issues, I could just rework the logic so that we're just creating a new `writer{}` and use it as a standalone.